cleaner.go

```diff
@@ -0,0 +1,362 @@
+	pullCount := 0
+	flushPulls := func() error {
+		result, err := pullChunk.Run()
+		if err != nil {
+			if err != mgo.ErrNotFound {
+				// not found is odd, but not considered fatal, all others are
+				return fmt.Errorf("error while updating documents: %v", err)
+			}
+		}
+		cleaner.stats.UpdatedDocCount += result.Matched
+		pullChunk = cleaner.config.Source.Bulk()
+		pullChunk.Unordered()
+		pullCount = 0
+		return nil
+	}
+	for _, doc := range cleaner.docsToProcess {
+		toPull := cleaner.findPullableTokens(doc.Queue, completedTokens)
+		if cleaner.removeIfEmpty && len(toPull) == len(doc.Queue) {
+			// this document can just be removed from the stash
+			cleaner.docIdsToRemove = append(cleaner.docIdsToRemove, doc.Id)
+		} else if len(toPull) > 0 {
+			// Note: (jam 2017-04-04) An observation, if it is legal to pull a token from one document,
+			// it is legal to pull it from all other documents. We could do bulk operations by using the
+			// union of all document ids and the union of all tokens to pull.
+			cleaner.stats.PulledTokenCount += len(toPull)
+			pull := bson.M{"$pullAll": bson.M{"txn-queue": toPull}}
+			pullChunk.Update(bson.M{"_id": doc.Id}, pull)
+			if pullCount >= maxBulkOps {
+				if err := flushPulls(); err != nil {
+					return err
+				}
+			}
+		}
+	}
+	if err := flushPulls(); err != nil {
+		return err
+	}
+	// We've handled these tokens and documents
+	cleaner.tokensToLookup = cleaner.tokensToLookup[:0]
+	cleaner.docsToProcess = cleaner.docsToProcess[:0]
+	return nil
+}
+
+// flushRemoveQueue ensures that all pending removals are flushed to the database.
+func (cleaner *collectionCleaner) flushRemoveQueue() error {
+	if len(cleaner.docIdsToRemove) == 0 {
+		return nil
+	}
+	remover := newBulkRemover(cleaner.config.Source)
+	for _, docId := range cleaner.docIdsToRemove {
+		if err := remover.remove(docId); err != nil {
+			return fmt.Errorf("failed while removing document %v from %q", docId, cleaner.config.Source.Name)
+		}
+	}
+	if err := remover.flush(); err != nil {
+		return fmt.Errorf("failed while removing documents from %q", cleaner.config.Source.Name)
+	}
+	cleaner.stats.RemovedCount += remover.removed
+	logger.Debugf("flushing %d documents removed %d (%d total)",
+		len(cleaner.docIdsToRemove), remover.removed, cleaner.stats.RemovedCount)
+	cleaner.docIdsToRemove = cleaner.docIdsToRemove[:0]
+	return nil
+}
+
+// Cleanup iterates the collection and ensures that all documents no longer reference completed transactions.
+func (cleaner *collectionCleaner) Cleanup() error {
+	// walk through all the documents and remove what we can, update what we cannot
+	logger.Debugf("cleaning up completed documents from %q", cleaner.config.Source.Name)
+	// If we delete documents while we iterate, it can cause the iterator to miss documents.
+	// So we do multiple passes on the database to make sure we catch everything.
+	var doc txnDocument
+	for iterCount := 0; iterCount < maxIterCount; iterCount++ {
```

Comment:
so again, this only triggers when we have found 100,000 documents that need to be removed (so only triggers on txns.stash), and it only queues it up to run another pass after it goes through all the documents.
So if we have, say, 1M documents that will be removed, and 100,000 that do not, we will pass through the first ~100,000, notice we need to remove them, remove them, keep going and find another 100,000, remove them, etc.
Once we've gone through everything, we'll then go back over the documents. We'll re-evaluate the 100,000 that didn't get removed, but then only trigger another pass if we find another 100,000 that *do* need to be removed.

I'll need to test again to see if we need to retry at all (since it might just be the order-preservation bug). Ultimately, I think this is still sane.

My particular testing said we could drop the max values before flush down if you want. I went ahead and dropped it to 50,000, which is probably under the "can I tell how much is used by this array" level.